src/kv/value.rs

```diff
@@ -635,6 +646,9 @@ pub trait Visit<'v> {
     /// more specific methods that aren't overridden.
     /// The `Value` may be formatted using its `fmt::Debug` or `fmt::Display` implementation,
     /// or serialized using its `sval::Value` or `serde::Serialize` implementation.
+    // REVIEW: Do we want seperate methods for sval and serde values? Otherwise
```

Comment:
That's a good question. The design here is that you shouldn't need separate methods, because you don't know what framework a user logged their structured data with, and as a consumer it doesn't matter to you. They pick the one that best suits them and you just pick the one that best suits you. `Value` will bridge them for you.

The drawback is it's really non-obvious how to work with it. You try to visit a `Value` to see what's in it and you get a method that gives you... a `Value`. This could just be an issue of documentation. The API itself doesn't give you a lot of clues about how to use it. That could be a reason to add separate methods, and examples alongside them on how to forward `visit_any` into them.

The only thing we'd need to worry about is whether `Visitor` needs to be object safe. I think it doesn't? In which case it's fine to add separate methods.

Reply:
> to determine if serde or sval is used

I think this is the part you shouldn't need to do in `visit_any`. In `std_logger`'s JSON format for instance, your [`visit_any`](https://github.com/Thomasdezeeuw/std-logger/blob/eff67f4c33b89c8debd325c1ad79724e6cd6af2c/src/format/json.rs#L169-L173) implementation could be:

```rust
fn visit_any(&mut self, value: kv::Value) -> Result<(), kv::Error> {
        Ok(serde_json::to_writer(&mut self.0, &value)?)
}
```

where `self.0` is `impl std::io::Write`. You don't need to know whether the value comes from `serde` or `sval`, or even need to override any of the other `Visitor` methods. You'll get the most accurate structured representation no matter what the value was captured with.

Reply:
> > to determine if serde or sval is used
> 
> I think this is the part you shouldn't need to do in `visit_any`. In `std_logger`'s JSON format for instance, your [`visit_any`](https://github.com/Thomasdezeeuw/std-logger/blob/eff67f4c33b89c8debd325c1ad79724e6cd6af2c/src/format/json.rs#L169-L173) implementation could be:
> 
> ```rust
> fn visit_any(&mut self, value: kv::Value) -> Result<(), kv::Error> {
>         Ok(serde_json::to_writer(&mut self.0, &value)?)
> }
> ```
> 
> where `self.0` is `impl std::io::Write`. You don't need to know whether the value comes from `serde` or `sval`, or even need to override any of the other `Visitor` methods. You'll get the most accurate structured representation no matter what the value was captured with.

So you're saying that `kv::Value` implements `serde::Deserialize` regardless of the source of the actual value (i.e. a `sval` value would be deserialised correctly?). And vice versa?

If so, it does essentially mean that any logging implementation that want to support structured value needs to use either `serde` or `sval`.

Reply:
> So you're saying that kv::Value implements serde::Deserialize regardless of the source of the actual value (i.e. a sval value would be deserialised correctly?). And vice versa?

That's essentially right :+1: It always implements `serde::Serialize` and the `Serialize` implementation will always use the least-lossy representation of the data you've captured.

> If so, it does essentially mean that any logging implementation that want to support structured value needs to use either serde or sval.

For simple values like numbers, strings, and booleans you don't need to. If you want to log complex values like maps or sequences then you'll need to use one of them. If you're happy to treat complex values as text then you can also always use `std::fmt`.

Reply:
https://github.com/rust-lang/log/pull/593